src/c_api/c_api.cc

```diff
@@ -502,12 +507,278 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,
+                   datacoltype = 3;
+                }
+                else if(wcscmp(stype,L"i1i")==0){
+                   datacoltype = 4;
+                }
+                else if(wcscmp(stype,L"i2i")==0){
+                   datacoltype = 5;
+                }
+                else if(wcscmp(stype,L"i8i")==0){
+                   datacoltype = 6;
+                }
+                else{
+                    fwprintf(stderr,L"Unknown type %s", stype);
+                    exit(1);
+                }
+//                fprintf(stderr,"chose datacoltype=%d for j=%d\n",datacoltype,j); fflush(stderr);
+            };
+//  ~datacol_struct() {}
+};
+
+// map dt stype string to C ctype for casting purposes
+// {'i1b': 'bool', 'i1i': 'int', 'i2i': 'int', 'i4i': 'int', 'i8i': 'int', 'f4r': 'float', 'f8r': 'float'}
+float get_dt_value(datacol_struct *d, int i)
+{
+//    fprintf(stderr,"using2 datacoltype=%d for i=%d\n",d->datacoltype,i); fflush(stderr);
+
+    // order of likelihood
+    switch(d->datacoltype){
+        case 0:
+          return static_cast<float>(d->datacol_float[i]);
+          break;
+        case 1:
+          return static_cast<float>(d->datacol_double[i]);
+          break;
+        case 2:
+          return static_cast<float>(d->datacol_bool[i]);
+          break;
+        case 3:
+          return static_cast<float>(d->datacol_int4[i]);
+          break;
+        case 4:
+          return static_cast<float>(d->datacol_int1[i]);
+          break;
+        case 5:
+          return static_cast<float>(d->datacol_int2[i]);
+          break;
+        case 6:
+          return static_cast<float>(d->datacol_int8[i]);
+          break;
+        default:
+          wprintf(L"Unknown type %s", d->stype);
+          fprintf(stderr,"Unknown datacoltype=%d\n",d->datacoltype);
+          fflush(stderr);
+          fflush(stdout);
+          exit(1);
+    }
+}
+
+
+bool is_dt_missing(datacol_struct *d, int i)
+{
+//    fprintf(stderr,"using1 datacoltype=%d for i=%d result=%g\n",d->datacoltype,i, d->datacol_double[i]); fflush(stderr);
+
+    // return false;
+    // fwprintf(stderr,L"stype = %s", stype);fflush(stderr);
+    // order of likelihood
+    if(d->datacoltype==0 && !std::isfinite(d->datacol_float[i])){ // GETNA<float>
+        return true;
+    }
+    else if(d->datacoltype==1 && !std::isfinite(d->datacol_double[i])){ // GETNA<double>
+        return true;
+    }
+    else if(d->datacoltype==2 && d->datacol_bool[i]==GETNA<bool>()){
+        return true;
+    }
+    else if(d->datacoltype==3 && d->datacol_int4[i]==GETNA<int32_t>()){
+        return true;
+    }
+    else if(d->datacoltype==4 && d->datacol_int1[i]==GETNA<int8_t>()){
+        return true;
+    }
+    else if(d->datacoltype==5 && d->datacol_int2[i]==GETNA<int16_t>()){
+        return true;
+    }
+    else if(d->datacoltype==6 && d->datacol_int8[i]==GETNA<int64_t>()){
+        return true;
+    }
+    else return false;
+}
```

Comment:
This all looks rather inefficient... `get_dt_value` and `is_dt_missing` are called within a tight loop, having lots of branching there kills the performance.
As a first step, I'd recommend the following: in class `datacol_struct` remove all members "datacol_*" and "datacoltype". Instead, create a hierarchy of classes `datacol_bool .. datacol_double` each deriving from common `datacol_struct` and exposing virtual methods `is_missing(int) -> bool` and `get_value(int) -> float`. Current constructor of `datacol_struct` would then become a factory method for instantiating one of these subclasses.

Reply:
I think switch statements and virtual functions are effectively equivalent AFAIK. Virtual functions incur a vtable function lookup that the cpu can use branch prediction for. https://stackoverflow.com/a/449967

Reply:
@st-pasha @RAMitchell After merging the is_missing and get_value, there is only one switch and it should be similar to any factory method.  I'll still ask you sometime how to go about the factory method, as it seems to defeat the purpose and require being explicit about which type I have.  The whole point of the class and switches is to input a string and get the correct casted value.  If you can do that with a factory in an easy way, I'll ask how.